inst/include/common/fims_math.hpp

```diff
@@ -220,6 +268,420 @@ inline const Type ad_max(const Type &a, const Type &b, Type C = 1e-5) {
+        }
+
+        static const T g = 7;
+        static const std::vector<T> coef = {
+            0.99999999999980993, 676.5203681218851,
+            -1259.1392167224028, 771.32342877765313,
+            -176.61502916214059, 12.507343278686905,
+            -0.13857109526572012, 9.9843695780195716e-6,
+            1.5056327351493116e-7
+        };
+
+        T x = coef[0];
+        for (size_t i = 1; i < coef.size(); ++i) {
+            x += coef[i] / (z + i - 1.0);
+        }
+
+        T t = z + g - 0.5;
+        return (fims_math::log(2.0 * M_PI) / 2.0) + ((z - 0.5) * fims_math::log(t)) - t + fims_math::log(x);
+    }
+
+    template<typename T>
+    fims::Vector<T> lgamma_ad(const fims::Vector<T>& v) {
+        fims::Vector<T> ret(v.size());
+        for (int i = 0; i < v.size(); i++) {
+            ret[i] = lgamma_ad(v[i]);
+        }
+        return ret;
+    }
+
+    template<class T>
+    T sum(const std::vector<T>& v) {
+        T ret = 0.0;
+        for (int i = 0; i < v.size(); i++) {
+            ret += v[i];
+        }
+        return ret;
+    }
+
+    template<class T>
+    T sum(const fims::Vector<T>& v) {
+        T ret = 0.0;
+        for (int i = 0; i < v.size(); i++) {
+            ret += v[i];
+        }
+        return ret;
+    }
+
+    /**
+     * Multinomial Probability Density function. p is internally normalized to sum 1.
+     * 
+     * @brief 
+     * 
+     * @param x
+     * @param p
+     * @param ret_log
+     * @return 
+     */
+    template<typename T>
+    T dmultinom(const fims::Vector<T>& x, fims::Vector<T> p, bool ret_log = false) {
+        if (x.size() != p.size()) {
+            throw std::invalid_argument("Size of x and p must be the same.");
+        }
+
+        T sum_x = fims_math::sum(x);
+        T log_factorial_sum_x = fims_math::lgamma_ad(sum_x + 1.0); // log(sum(x)!)
+
+        T log_factorial_x = 0;
+        T sum_x_log_p = 0;
+
+        for (size_t i = 0; i < x.size(); ++i) {
+            if (p[i] <= 0 || p[i] >= 1) {
+                throw std::invalid_argument("Probabilities must be between 0 and 1, exclusive.");
+            }
+            log_factorial_x += fims_math::lgamma_ad(x[i] + 1); // log(x_i!)
+            sum_x_log_p += x[i] * fims_math::log(p[i]); // x_i * log(p_i)
+        }
+        T ret = log_factorial_sum_x - log_factorial_x + sum_x_log_p;
+
+        if (ret_log) {
+            return ret;
+        } else {
+            return fims_math::exp(ret);
+        }
+    }
+
+    template <typename T>
+    fims::Vector<int> rmultinom(const fims::Vector<T>& prob) {
+        std::random_device rd;
+        std::mt19937 gen(rd());
+        std::discrete_distribution<> dist(prob.begin(), prob.end());
+
+        size_t n = prob.size();
+        fims::Vector<int> result(prob.size(), 0);
+        for (int i = 0; i < n; ++i) {
+            int index = dist(gen);
+            ++result[index];
+        }
+
+        return result;
+    }
+
+
+
```

Comment:
The task of defining our own distributions was scoped out of the current milestone so we haven't developed a design doc yet for best practices (ie. writing functions in logspace, whether or not to approximate the lgamma and take its derivative or approximate the derivative of the lgamma, etc.) and how we want to organize the code. Can we save these functions elsewhere until we've gone through these additional steps?

Reply:
For right now, if we are going to leave them in, I think that they should be returning the same values as R and TMB, which means that they need to be in log space. But, I don't know enough about the lgamma to know if it is written correctly. Each of these should be flagged with a comment though that they are not currently used and the code is beta.